synapse/lib/cmdr.py

```diff
@@ -9,36 +20,595 @@
+            ('cliserver:banner', {'type': 'str', 'defval': '',
+                                  'doc': 'Banner displayed to user upon on connection. It is sent prior to '
+                                         'postClientInit being called.'})
+        )
+        return confdefs
+
+    def _finiServer(self):
+        # Tear down our CLI objects via their .sock objects so they are
+        # properly cleaned up and we only have to care about defining fini
+        # routines in a single place
+        clis = list(self.sessions.items())
+        for iden, cli in clis:
+            cli.sock.fini()
+
+    def _broadcastMesg(self, mesg, addnl=True):
+        '''
+        Broadcast a message to all connected CLI clients.
+
+        Args:
+            mesg (str): Message to send.
+            addnl (bool): Add a newline to the message.
+
+        Notes:
+            This method is private so that it may not be exposed to remote
+            users via Telepath. CliServer implementors may choose to use it
+            in their implementations.
+
+        Returns:
+            None
+        '''
+        self.outp.printf(mesg, addnl=addnl)
+
+    def postServerInit(self):
+        '''
+        This is executed at the end of the CliServer __init__ method.
+
+        Implementors may override this to provide additional functionality..
+        '''
+        pass
+
+    def getCmdrSession(self, proxyiden):
+        '''
+        Establissh a remote CLI session with the current socket connection.
+
+        Args:
+            proxyiden (str): An iden for the proxy.
+
+        Examples:
+            Connect to the CliServer and get a new session, then listen for print events specific for the session:
+
+                prox = s_telepath.openurl('tcp://cliserver.com:8675/mycli')
+                session_iden = prox.getCmdrSession(guid())
+                prox.on('syn:cliserver:outp:%s' % session_iden, some_print_func)
+
+        Notes:
+            This method is typically called by the CliProxy classs directly
+            and most users do not need to ever call this manually.
+
+        Returns:
+            str: A session iden.
+
+        Raises:
+            SessAlreadyExists: If the current connection already has an existing CLI session.
+            BadTypeValu: If the proxyiden is not a guid.
+        '''
+
+        sock = s_scope.get('sock')
+        user = s_scope.get('user')
+
+        cliden = sock.get('cli:iden')
+        if cliden is not None:
+            raise s_common.SessAlreadyExists(mesg='sock already has a cli iden')
+        if not s_common.isguid(proxyiden):
+            raise s_common.BadTypeValu(valu=proxyiden,
+                                       mesg='proxyiden must be a guid')
+
+        # Make the proxy iden locally available
+        s_scope.set('proxy:iden', proxyiden)
+        s_scope.set('server', self)
+
+        # Make the per-session iden / outp
```

Comment:
We probably want to throw a lock around the session iden creation and ensure we're not stamping the same iden in twice